### PR TITLE
Open datasets

### PR DESCRIPTION
**Pull request by NicolasLegendre1**

### CODE REVIEW

**Comment by fredericpoitevin** [comment]
Thanks @NicolasLegendre1 for contributing this!

I think we should have "data/input" files in a directory distinct from the code. I'd suggest to move the `.npy` file to `data/` please.

File: cryo/dataset.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import functools
+from PIL import Image
+
+CUDA = torch.cuda.is_available()
+
+KWARGS = {'num_workers': 1, 'pin_memory': True} if CUDA else {}
+
+
+def open_dataset(path, new_size, is_3d):
+    """Open datasets and processes data in ordor to make tensors.
```
Comment (fredericpoitevin) [inline cryo/dataset.py:18]: NIT: "in order to"

File: cryo/dataset.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    Parameters
+    ----------
+    path : string,
```
Comment (fredericpoitevin) [inline cryo/dataset.py:22]: What file types are accepted? What does the function expects those files to contain? Please tell us more here.

File: cryo/dataset.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        Path.
+    new_size : int.
```
Comment (fredericpoitevin) [inline cryo/dataset.py:24]: Please tell us what `size` means.

File: cryo/dataset.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    is_3d : boolean
+        If 2d or 3d.
+
+    Returns
+    -------
+    tensor
+        Images in black and white.
```
Comment (fredericpoitevin) [inline cryo/dataset.py:31]: NIT: greyscale images.

File: cryo/dataset.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    ----------
+    path : string,
+        Path.
+    new_size : int.
+    is_3d : boolean
+        If 2d or 3d.
+
+    Returns
+    -------
+    tensor
```
Comment (fredericpoitevin) [inline cryo/dataset.py:30]: If I understand the code correctly, this tensor is specifically a `torch` tensor. Would it be readable by other backends such as `tensorflow` or `numpy`? If not, please mention it here.

File: cryo/dataset.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        dataset = torch.Tensor(new_dataset)
+        dataset = normalization_linear(dataset)
+        if len(img_shape) != 4:
+            dataset = dataset.reshape(
+                (img_shape[0], 1, new_size, new_size))
+    return dataset
+
+
+def normalization_linear(dataset):
+    """Normalize a tensor.
```
Comment (fredericpoitevin) [inline cryo/dataset.py:66]: same comment about backend: is this tensor specific to pytorch?

Comment (fredericpoitevin) [reply cryo/dataset.py:66]: Idea: if those functions are torch specific, I'd suggest to have the name say it explicitely, e.g.: `normalization_linear_torch()`

File: cryo/dataset.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                    dataset[i][0]).resize([new_size, new_size])))
+        dataset = torch.Tensor(new_dataset)
+        dataset = normalization_linear(dataset)
+        if len(img_shape) != 4:
+            dataset = dataset.reshape(
+                (img_shape[0], 1, new_size, new_size))
+    return dataset
+
+
+def normalization_linear(dataset):
```
Comment (fredericpoitevin) [inline cryo/dataset.py:65]: How about `normalize(dataset, scale='linear')` ?
That would open the function to maybe have more ways to normalize?

File: cryo/dataset.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        return tuple(obj['items'])
+    return obj
+
+
+def load_parameters(path):
+    """Load metadata for the VAE.
+
+    Parameters
+    ----------
+    path : string,
```
Comment (fredericpoitevin) [inline cryo/dataset.py:145]: ditto

File: cryo/test_dataset.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+"""Test dataset."""
+import dataset as ds
+import torch
+import numpy as np
+
+
+class TestDataset:
```
Comment (fredericpoitevin) [inline cryo/test_dataset.py:7]: Fantastic @NicolasLegendre1 ! :D

**Review by fredericpoitevin — CHANGES_REQUESTED** [review]
Pending Nina's approval, I think we can merge as soon as you move the files to `data/` and address the very few comments I made. Thanks @NicolasLegendre1

File: cryo/dataset.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                (img_shape[0], 1, new_size, new_size))
+    return dataset
+
+
+def normalization_linear(dataset):
+    """Normalize a tensor.
+
+    Parameters
+    ----------
+    dataset : tensor,
```
Comment (ninamiolane) [inline cryo/dataset.py:70]: NIT: no , at the end here

Comment (ninamiolane) [reply cryo/dataset.py:70]: (and same thing evrywhere else in the code)

File: cryo/test_dataset.py (hunk 9 above)
Comment (ninamiolane) [reply cryo/test_dataset.py:7]: Agree!! 🎉

**Review by ninamiolane — CHANGES_REQUESTED** [review]
Very nice!! Just a few thing to adapt.

**Comment by ninamiolane** [comment]
I'm not sure that the `.json` files should go to `data/` since they are parameters and not data?

As for the `data_test.npy`, these are not really data, but toy data (just here to run the tests), so they could stay in somewhere in the `tests/` folder?

The real data cannot be stored on GitHub (way too big), thus we will need another public repository to store them and make them available to our users.

File: cryo/test_dataset.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
+"""Test dataset."""
```
[the remark this answers is not in the document]
Comment (fredericpoitevin) [reply cryo/test_dataset.py:1]: Right, I missed this. It's important.

I would suggest the following:
```
ioSPI/
L____ iotools/
         L______ atomic_models.py
         L______ cryo_dataset.py.     <==== (currently cryo/dataset.py)
tests/
L____ test_iotools/
         L___________ test_atomic_models.py
         L___________ test_cryo_dataset.py            <==== (currently cryo/test_dataset.py)
         L___________ data_test_cryo_dataset.npy <==== (currently cryo/data_test.npy) 
```

A few notes:
- I haven't homogenized `iospi` ==> `ioSPI` everywhere yet. Should be done asap.
- I think we want `cryo/vae_parameters.json` to go to `data/`

**Comment by fredericpoitevin** [comment]
> I'm not sure that the `.json` files should go to `data/` since they are parameters and not data?
So we want to have it under `ioSPI/iotools/`?

> As for the `data_test.npy`, these are not really data, but toy data (just here to run the tests), so they could stay in somewhere in the `tests/` folder?
OK makes sense.

> The real data cannot be stored on GitHub (way too big), thus we will need another public repository to store them and make them available to our users.
Agreed.
